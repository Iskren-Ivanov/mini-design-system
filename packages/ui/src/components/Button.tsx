import React from 'react';
import { cn } from '../utils/cn';

export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: 'primary' | 'secondary' | 'outline' | 'ghost';
  size?: 'sm' | 'md' | 'lg';
  children: React.ReactNode;
}

const buttonVariants = {
  primary: 'bg-[var(--color-primary)] text-white hover:bg-[var(--color-primary-hover)] focus:ring-[var(--color-primary)] shadow-[var(--shadow-sm)]',
  secondary: 'bg-[var(--color-bg-tertiary)] text-[var(--color-text-primary)] hover:bg-[var(--color-border-secondary)] focus:ring-[var(--color-border-secondary)] shadow-[var(--shadow-sm)]',
  outline: 'border-2 border-[var(--color-primary)] bg-transparent text-[var(--color-primary)] hover:bg-[var(--color-primary-light)] focus:ring-[var(--color-primary)]',
  ghost: 'bg-transparent text-[var(--color-text-secondary)] hover:bg-[var(--color-bg-tertiary)] focus:ring-[var(--color-border-secondary)]',
};

const buttonSizes = {
  sm: 'px-3 py-1.5 text-sm font-medium',
  md: 'px-4 py-2 text-base font-semibold',
  lg: 'px-6 py-3 text-lg font-bold',
};

export function Button({
  variant = 'primary',
  size = 'md',
  className,
  children,
  disabled,
  ...props
}: ButtonProps) {
  return (
    <button
      className={cn(
        'inline-flex items-center justify-center rounded-md font-medium transition-colors',
        'focus:outline-none focus:ring-2 focus:ring-offset-2',
        'disabled:opacity-50 disabled:pointer-events-none',
        buttonVariants[variant],
        buttonSizes[size],
        className
      )}
      disabled={disabled}
      {...props}
    >
      {children}
    </button>
  );
}
