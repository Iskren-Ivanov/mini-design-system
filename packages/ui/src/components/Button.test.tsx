import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { Button } from './Button';

describe('Button', () => {
  it('renders the button with default variant and medium size', () => {
    render(<Button>Click Me</Button>);
    const button = screen.getByRole('button', { name: /click me/i });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass('bg-[var(--color-primary)]');
    expect(button).toHaveClass('px-4 py-2 text-base');
  });

  it('renders with primary variant', () => {
    render(<Button variant='primary'>Primary Button</Button>);
    const button = screen.getByRole('button', { name: /primary button/i });
    expect(button).toHaveClass('bg-[var(--color-primary)]');
  });

  it('renders with secondary variant', () => {
    render(<Button variant='secondary'>Secondary Button</Button>);
    const button = screen.getByRole('button', { name: /secondary button/i });
    expect(button).toHaveClass('bg-[var(--color-bg-tertiary)]');
  });

  it('renders with outline variant', () => {
    render(<Button variant='outline'>Outline Button</Button>);
    const button = screen.getByRole('button', { name: /outline button/i });
    expect(button).toHaveClass('border-2 border-[var(--color-primary)]');
  });

  it('renders with ghost variant', () => {
    render(<Button variant='ghost'>Ghost Button</Button>);
    const button = screen.getByRole('button', { name: /ghost button/i });
    expect(button).toHaveClass('bg-transparent');
  });

  it('renders with small size', () => {
    render(<Button size='sm'>Small Button</Button>);
    const button = screen.getByRole('button', { name: /small button/i });
    expect(button).toHaveClass('px-3 py-1.5 text-sm');
  });

  it('renders with large size', () => {
    render(<Button size='lg'>Large Button</Button>);
    const button = screen.getByRole('button', { name: /large button/i });
    expect(button).toHaveClass('px-6 py-3 text-lg');
  });

  it('handles click events', () => {
    const handleClick = vi.fn();
    render(<Button onClick={handleClick}>Clickable</Button>);
    const button = screen.getByRole('button', { name: /clickable/i });
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('is disabled when disabled prop is true', () => {
    render(<Button disabled>Disabled Button</Button>);
    const button = screen.getByRole('button', { name: /disabled button/i });
    expect(button).toBeDisabled();
    expect(button).toHaveClass('disabled:opacity-50');
  });

  it('applies custom className', () => {
    render(<Button className='custom-class'>Custom Button</Button>);
    const button = screen.getByRole('button', { name: /custom button/i });
    expect(button).toHaveClass('custom-class');
  });

  // Snapshot tests
  it('matches snapshot for primary variant', () => {
    const { asFragment } = render(<Button variant='primary'>Primary Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot for secondary variant', () => {
    const { asFragment } = render(<Button variant='secondary'>Secondary Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot for outline variant', () => {
    const { asFragment } = render(<Button variant='outline'>Outline Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot for ghost variant', () => {
    const { asFragment } = render(<Button variant='ghost'>Ghost Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot for disabled state', () => {
    const { asFragment } = render(<Button disabled>Disabled Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

});
