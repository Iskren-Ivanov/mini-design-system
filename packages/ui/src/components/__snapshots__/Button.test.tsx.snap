// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot for disabled state 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center rounded-md transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none bg-[var(--color-primary)] text-white hover:bg-[var(--color-primary-hover)] focus:ring-[var(--color-primary)] shadow-[var(--shadow-sm)] px-4 py-2 text-base font-semibold"
    disabled=""
  >
    Disabled Button
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot for ghost variant 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center rounded-md transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none bg-transparent text-[var(--color-text-secondary)] hover:bg-[var(--color-bg-tertiary)] focus:ring-[var(--color-border-secondary)] px-4 py-2 text-base font-semibold"
  >
    Ghost Button
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot for outline variant 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center rounded-md transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none border-2 border-[var(--color-primary)] bg-transparent text-[var(--color-primary)] hover:bg-[var(--color-primary-light)] focus:ring-[var(--color-primary)] px-4 py-2 text-base font-semibold"
  >
    Outline Button
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot for primary variant 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center rounded-md transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none bg-[var(--color-primary)] text-white hover:bg-[var(--color-primary-hover)] focus:ring-[var(--color-primary)] shadow-[var(--shadow-sm)] px-4 py-2 text-base font-semibold"
  >
    Primary Button
  </button>
</DocumentFragment>
`;

exports[`Button > matches snapshot for secondary variant 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center rounded-md transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none bg-[var(--color-bg-tertiary)] text-[var(--color-text-primary)] hover:bg-[var(--color-border-secondary)] focus:ring-[var(--color-border-secondary)] shadow-[var(--shadow-sm)] px-4 py-2 text-base font-semibold"
  >
    Secondary Button
  </button>
</DocumentFragment>
`;
